zone/zonedb.cpp

```diff
@@ -1058,18 +1059,35 @@ bool ZoneDatabase::SaveCharacterDisc(uint32 character_id, uint32 slot_id, uint32
 	return true;
 }
 
-bool ZoneDatabase::SaveCharacterTribute(uint32 character_id, PlayerProfile_Struct* pp){
-	std::string query = StringFormat("DELETE FROM `character_tribute` WHERE `id` = %u", character_id);
-	QueryDatabase(query);
-	/* Save Tributes only if we have values... */
-	for (int i = 0; i < EQ::invtype::TRIBUTE_SIZE; i++){
-		if (pp->tributes[i].tribute >= 0 && pp->tributes[i].tribute != TRIBUTE_NONE){
-			std::string query = StringFormat("REPLACE INTO `character_tribute` (id, tier, tribute) VALUES (%u, %u, %u)", character_id, pp->tributes[i].tier, pp->tributes[i].tribute);
-			QueryDatabase(query);
-			LogDebug("ZoneDatabase::SaveCharacterTribute for character ID: [{}], tier:[{}] tribute:[{}] done", character_id, pp->tributes[i].tier, pp->tributes[i].tribute);
+void ZoneDatabase::SaveCharacterTribute(Client* c)
+{
+	std::vector<CharacterTributeRepository::CharacterTribute> tributes = {};
+	CharacterTributeRepository::CharacterTribute tribute = {};
+
+	uint32 tribute_count = 0;
+	for (auto& t : c->GetPP().tributes) {
+		if (t.tribute != TRIBUTE_NONE) {
+			tribute_count++;
 		}
 	}
-	return true;
+
+	tributes.reserve(tribute_count);
+
+	for (auto& t : c->GetPP().tributes) {
+		if (t.tribute != TRIBUTE_NONE) {
+			tribute.character_id = c->CharacterID();
+			tribute.tier         = t.tier;
+			tribute.tribute      = t.tribute;
+
+			tributes.emplace_back(tribute);
+		}
+	}
+
+	if (tribute_count > 0) {
+		CharacterTributeRepository::DeleteWhere(database, fmt::format("character_id = {}", c->CharacterID()));
```

Comment:
We should probably pull this before `if (tribute_count > 0) {` because we could end up leaving stale entries behind and re-loading them

Reply:
I’ll move it this afternoon. Was following the recent bulk update stuff you did as an example.

Reply:
Understood, situation-ally my code was wrong and you saw it was fixed in the recent PR